File: happy_number.java
// Happy Number
// 19 is a happy number

// 1^2 + 9^2 = 82
// 8^2 + 2^2 = 68
// 6^2 + 8^2 = 100
// 1^2 + 0^2 + 0^2 = 1

public class Solution {
    /**
     * @param n an integer
     * @return true if this is a happy number or false
     */
    public boolean isHappy(int n) {
        // Write your code here
        HashSet<Integer> set = new HashSet<Integer>();
        while(!set.contains(n)){
            set.add(n);
            n = sum(getDigits(n));
            if(n == 1){
                return true;
            }

        }
        return false;
    }
    public int sum(int[] arr){
        int sum = 0;
        for(int i = 0; i < arr.length; i++){
            sum += arr[i]*arr[i];
        }
        return sum;
    }
    
    public int[] getDigits(int n){
        String s = String.valueOf(n);
        int[] res = new int[s.length()];
        int i = 0;
        while(n > 0){
            int m = n%10;
            res[i++] = m;
            n /= 10;
        }
        return res;
    }

}

File: happyNumber.js
// Write an algorithm to determine if a number is "happy".

// A happy number is a number defined by the following process: Starting with any positive integer, replace the number by the sum of the squares of its digits, and repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1. Those numbers for which this process ends in 1 are happy numbers.

// Example: 19 is a happy number

// 12 + 92 = 82
// 82 + 22 = 68
// 62 + 82 = 100
// 12 + 02 + 02 = 1

/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
    var map = {};
    while(map[n] === undefined){
        map[n] = 1;
        n = getSum(getDigits(n));
    }
    return n === 1;
};

var getDigits = function(n){
    var res = [];
    while(n > 0){
        res.push(n % 10);
        n = parseInt(n/10);
    }
    return res;
}

var getSum = function(arr){
    var sum = 0;
    for(var x in arr){
        sum+=arr[x]*arr[x];
    }
    return sum;
}